components/docker-auth/cmd/authn/authentication.go

```diff
@@ -105,15 +132,129 @@ func main() {
 	tokenValidity, err := validateToken(token, certificateInUse)
 	if err != nil {
 		glog.Errorf("Token is not valid - %s", err)
-		glogFlushAndExit(extension.ErrorExitCode)
+		extension.Exit(extension.ErrorExitCode)
 	}
 	glog.Info("signature verified")
 
 	if tokenValidity {
 		glog.Info("user successfully authenticated")
-		glogFlushAndExit(extension.SuccessExitCode)
+		extension.Exit(extension.SuccessExitCode)
 	} else {
 		glog.Error("authentication failed")
-		glogFlushAndExit(extension.ErrorExitCode)
+		extension.Exit(extension.ErrorExitCode)
+	}
+}
+
+// isJWT checks whether the token is jwt token or access token.
+func isJWT(conf extension.Conf) bool {
+	isJWTEnv := os.Getenv("IS_JWT")
+	var isJWT bool
+	if len(isJWTEnv) == 0 {
+		isJWT = conf.AuthConf.IsJWT
+	} else {
+		if isJWTEnv == "true" {
+			isJWT = true
+		} else if isJWTEnv == "false" {
+			isJWT = false
+		} else {
+			glog.Error("Wrong environment value given. The value should be either true or false")
+			extension.Exit(extension.ErrorExitCode)
+		}
+	}
+	return isJWT
+}
+
+// validateAccessToken is used to introspect the access token
+func validateAccessToken(token string, providedUsername string, conf extension.Conf) bool {
+	idpHost, idpPort := resolveISHostAndPort(conf)
+	url := "https://" + idpHost + ":" + idpPort + "/oauth2/introspect"
+	payload := strings.NewReader("token=" + token)
+	req, err := http.NewRequest("POST", url, payload)
+	if err != nil {
+		glog.Error("Error creating new request to the introspection endpoint: ", err)
+		extension.Exit(extension.ErrorExitCode)
+	}
+
+	username, password := resolveCredentials(conf)
+	req.SetBasicAuth(username, password)
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		glog.Error("Error sending the request to the introspection endpoint: ", err)
+		extension.Exit(extension.ErrorExitCode)
+	}
+	defer res.Body.Close()
+	body, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		glog.Error("Error reading the response from introspection endpoint: ", err)
+		extension.Exit(extension.ErrorExitCode)
+	}
+	var result map[string]interface{}
+	err = json.Unmarshal([]byte(string(body)), &result)
+	if err != nil {
+		glog.Error("Error un marshalling the json: ", err)
+		extension.Exit(extension.ErrorExitCode)
+	}
+	isActive, ok := (result["active"]).(bool)
+	if !ok {
+		glog.Error("Error casting active to boolean. This may be due to a invalid token")
+		extension.Exit(extension.ErrorExitCode)
+	}
+	isExpired := isExpired(result["exp"])
+	isValidUser := isValidUser(result["username"], providedUsername)
+	return isExpired && isActive && isValidUser
+}
+
+// resolves the IS host and port from the environment variables.
+// If the environment is not set the port and host will be resolved through the config file.
+func resolveISHostAndPort(conf extension.Conf) (string, string) {
```

Comment:
Why don't we just have a single method which loads the config with default values for empty fields ?
Having too many methods like this seems to be complicating the config loading too much.

Then when accessing the values of the config you can directly access the field of the struct without having a separate method to do so.